src/main/java/org/openjfx/JavaFXBaseMojo.java

```
@@ -274,11 +283,32 @@ void preparePaths(Path jdkHome) throws MojoExecutionException {
                     }
                 });
             }
-
         } catch (Exception e) {
             getLog().warn(e.getMessage());
         }
 
+        if (runtimePathOption == MODULEPATH) {
+            getLog().debug(runtimePathOption + " runtimePathOption set by user. Moving all jars to modulepath.");
+            if (moduleDescriptor == null) {
+                // target/classes should still be on classpath
+                final List<String> classpathJars = classpathElements.stream()
+                        .filter(ce -> ce.endsWith(".jar"))
+                        .collect(Collectors.toList());
+                modulepathElements.addAll(classpathJars);
+                classpathElements.removeAll(classpathJars);
+            } else {
+                modulepathElements.addAll(classpathElements);
+                classpathElements.clear();
+            }
+        } else if (runtimePathOption == CLASSPATH) {
+            getLog().debug(runtimePathOption + " runtimePathOption set by user. Moving all jars to classpath.");
+            classpathElements.addAll(modulepathElements);
+            modulepathElements.clear();
+            if (doesExtendFXApplication(mainClass)) {
```

if mainClass is in the form moduleName/packge.name.MainClass, you'll need to strip out the class name, otherwise `Class.forName(mainClass)` will fail

---

Should we fail-fast if mainClass contains module-name and CLASSPATH runtime option is used together?

---

That goes in the line of my opinion of failing fast if there is a module-descriptor in the first place... Apply the same logic in both cases.

---

I still do not want to fail when module-descriptor is present :)

---

What's the difference of having a module-descriptor and not failing if class is not prefixed with moduleName, and failing when having a module-descriptor and class is prefixed with the moduleName?

---

You do have a point here. It is better to ignore the moduleName from class and log something like "Main module found in <mainClass> with runtimePathOption set as CLASSPATH. Module name will be ignored."